### ISSUE TITLE
GitGutter isn't working with encoding for shift jis (because ConvertToUTF8 marks view as scratch)

### ISSUE THREAD

**Issue by yuki1318** [jisaacks/GitGutter#425]

**Comment by deathaxe** [comment]
1. Please open ST's console 
2. Execute the following command: `view.encoding()`
3. Post the command's output here, please

Do you use a third party package such as _ConvertToUTF8_ to add support for this encoding or does it work out of the box?

**Comment by yuki1318** [comment]
Yes, I do.
I use ConvertToUTF8 for shift_jis to utf-8.

>>> view.encoding()
'UTF-8'

Source code encoding is shift_jis.
What should I do？

**Comment by deathaxe** [comment]
Can you post the result of `self.view.settings().get('origin_encoding')` command, too, if you have one of your source files open? It should return something like `shift_jis`


GitGutter reads the files from your repository as is without any charset convertion. This means the file your view is compared against is still `shift_jis`.

The content of the view was converted to utf-8. GitGutter uses the value of `self.view.settings().get('origin_encoding')` in case of installed ConvertToUTF8 package to get the original encoding of the view. 

I guess the return value is either wrong or python fails to encode the view content as `shift_jis` which causes a fallback to `utf-8`.

As a result the files passed to git diff are different and therefore anything is marked modified.

**Comment by yuki1318** [comment]
>>> view.settings().get('origin_encoding')
'Shift_JIS'

GitGutter ver 1.4.0 was working with encoding for shift jis, But after 1.5.0 has stopped working
Please tell me the solution strategy.

**Comment by deathaxe** [comment]
GitGutter 1.4.0 used `git show` to read the file content from repository. This command does not support _smudge filters_ (see issue #74). Therefore 1.5.0 changed to `git archive` command, which introduced issue #409 as this commad supports _smudge_ but ignores some files declared in `.gitattributes`. The current version therefore uses `git cat-file --filters` to read file content as this fixes both mentioned issues.

I guess the `git cat-file --filters` command being used at the moment does no text decoding and therefore the file is saved to disk with Shift_JIS encoding. GitGutter tries to write the encoded view content to disk using the value from either `view.settings().get('origin_encoding')` or `view.encoding()` but this obviously fails as python doesn't know Shift_JIS out of the box.

The former command `git show` seems to output the file content as utf-8 which is the default for all high level outputs from git. Therefore no one ever saw GitGutter failing in encoding the view content back to the `origin_encoding`.

This is the third issue besides #74 and #409 related to reading files from git. I don't see a solution how modify GitGutter to fix all three issues at the moment. 

_GitGutter supports all file encodings supported by python/Sublime Text out of the box._

ConvertToUTF8 is a big hack to try to work around missing file encodings. GitGutter can't make use of it as it doesn't publish any API to convert files in the background.

**Stupid question from someone not dealing with eastern languages too often:**
Why do you use a local (lagecy) file encoding to store files in a git repository? From my point of view ALL files on each computer should use utf-8 or utf-16 nowadays. If you would setup git to store files using utf-8, everything would work fine.

**Comment by deathaxe** [comment]
Two questions:

1. Which Sublime Text version do you use?

2. Can you create a little repository on github with a source/text file encoded in Shift_JIS which I could pull and use for debugging?

The python3 module of ST3 seems to be able to encode/decode 'Shift_JIS' out of the box. Using codecs of ST2 generates some totally different output. Without any reference I can't find out what's wrong.

**Comment by yuki1318** [comment]
Thank you for being kind enough to explain

>>Stupid question from someone not dealing with eastern languages too often:
Why do you use a local (lagecy) file encoding to store files in a git repository? From my point of view ALL files on each computer should use utf-8 or utf-16 nowadays. If you would setup git to store files using utf-8, everything would work fine.

I think so, too. But there are some problems, Older projects are made with shiftjis and compilers is not support utf-8. So I need to use shift_jis.


>> 1.Which Sublime Text version do you use? (You can also use GitGutter: Support Info from command pallet)
- Sublime Text 3126
- Platform: windows
- Arch: x64
- GitGutter 586dc0c
- Install via PC: False
- git version 2.10.1.windows.1
- mdpopups 2.0.0
- markdown 2.6.6
- pygments 2.1a0
- jinja2 2.8

>> 2.Can you create a little repository on github with a source/text file encoded in Shift_JIS which I could pull and use for debugging?

OK.
I create it. (https://github.com/yuki1318/TestRepos)

**Comment by deathaxe** [comment]
Ok, I checked out your repo, installed ConvertToUtf8 and opened your main.c

_The good news:_ The reason is much simpler then I original thought.

_The bad news:_ It's not GitGutter which causes this issue but ConvertToUtf8 and/or ST.

GitGutter just keeps disabled for the file as the view is marked as scratch by `view.set_scratch(True)`.

_Scratch views are most likely views with content not being saved to disk so don't need to be diffed. Terminal packages use scratch views as an example. New files not yet saved are marked as scratched, too._

I would primarily blame _ConvertToUtf8_ not to reset the scratch flag after transcoding the file, but I found ST 3136 working properly with the same setup. Maybe the core devs added some extra checks to reset the scratch flag.

On ST 3126 I just called `view.set_scratch(False)` in console and voila ... GitGutter works properly. Git check's out files as is with Shift_JIS encoding and the view content is encoded correctly to Shift_JIS. Everything works fine.

_Note_

I added a number of checks to keep GitGutter calm down for views most likely not meant for normal use. This helps to save some resources and keep ST responsive or avoids edge cases.
